bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,376 @@
+                        prepare_get(vec![(val, path.to_string())], app_args.keysize)
+                    }
+                    Some(KeyPattern::Range(_)) | Some(KeyPattern::Multiple(_)) => {
+                        return error!("Multiple keys are not allowed without pattern")
+                    }
+                    None => unreachable!(),
+                },
+                None => unreachable!(),
+            };
+            for kn in keys_names {
+                info!(
+                    "GET key:\"{:?}\" from  \"{}\" to file \"{}\"",
+                    kn.key, addr, &kn.name
+                );
+                get(kn.key, &kn.name, addr.clone()).await; // i dont like this clone()...
             }
-            "get" => {
-                let addr = sub_mathes
-                    .value_of("uri")
-                    .expect("has default value")
-                    .parse()
-                    .expect("wrong format of url");
-                info!("GET key:\"{:?}\" from  \"{}\"", key, addr);
-                get(key, addr).await;
+        }
+        EXISTS_SC => {
+            match app_args.key_pattern {
+                Some(KeyPattern::Single(val)) => {
+                    let key = get_key_value(val, app_args.keysize);
+                    exist(vec![key], addr).await;
+                }
+                Some(KeyPattern::Range(range)) => {
+                    // FIXME: empty response when range is descending (14-1)
+                    let keysize = app_args.keysize;
+                    let keys = range.map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
+                Some(KeyPattern::Multiple(vec)) => {
+                    let keysize = app_args.keysize;
+                    let keys = vec.into_iter().map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
+                None => unreachable!()
             }
-            _ => {}
         }
+        _ => unreachable!(),
     }
 }
 
-async fn put(key: Vec<u8>, size: usize, addr: Uri) {
-    let mut client = BobApiClient::connect(addr).await.unwrap();
+async fn prepare_put_from_pattern(re_path: &Regex, dir: &PathBuf, key_size: usize) -> Vec<KeyName> {
+    let mut dir_iter = read_dir(dir).await.unwrap();
 
-    let timestamp = SystemTime::now()
-        .duration_since(UNIX_EPOCH)
-        .expect("msg: &str")
-        .as_secs();
-    let meta = BlobMeta { timestamp };
-    let blob = Blob {
-        data: vec![1; size],
-        meta: Some(meta),
-    };
-    let message = PutRequest {
-        key: Some(BlobKey { key }),
-        data: Some(blob),
-        options: None,
-    };
-    let put_req = Request::new(message);
+    let mut keys_names = Vec::new();
+    while let Some(entry) = dir_iter.next_entry().await.unwrap() {
+        let file = entry.path();
+        if !file.is_dir() && re_path.is_match(file.to_str().unwrap()) {
```

Comment:
IMO It's better to use `captures` single time and just check if it's `Some` instead of doing regex search twice

Reply:
Done: [d721ce3](https://github.com/qoollo/bob/pull/562/commits/d721ce36e752eb5d25ed17726088fea578ed5550)